tiledb/sm/compressors/dict_compressor.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file   dict_compressor.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the dictionary compression class for strings.
+ */
+
+#ifndef TILEDB_DICT_COMPRESSOR_H
+#define TILEDB_DICT_COMPRESSOR_H
+
+#include "tiledb/common/common.h"
+#include "tiledb/sm/misc/endian.h"
+
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+/** Handles dictionary compression/decompression of variable sized strings. */
+class DictEncoding {
+ public:
+  /**
+   * Compress variable sized strings into dictionary encoded format
+   *
+   * @tparam T Type of integer that can fit the word ids
+   * @param input Input in form of a memory contiguous sequence of
+   * strings
+   * @param output Dictionary-encoded output in ids of type T. Memory is
+   * allocated and owned by the caller
+   * @return A dictionary in the form of a vector of strings, where indices are
+   * the word_ids of each word
+   */
+  template <class T>
+  static std::vector<std::string_view> compress(
+      const span<std::string_view> input, span<T> output) {
+    if (input.empty() || output.empty()) {
+      return {};
+    }
+
+    // We use string_views in the dictionary, which point to input strings. This
+    // means that input should not be freed before dictionary data is consumed
+    // TBD: Is this acceptable, or do we use dict of strings - more expensive to
```

Comment:
I think this optimization is nice and acceptable. Please highlight the potential lifetime issue in the function documentation.

Reply:
done